packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -0,0 +1,3450 @@
+#include <KokkosBatched_LU_Decl.hpp>
+#include <KokkosBatched_LU_Serial_Impl.hpp>
+#include <KokkosBatched_LU_Team_Impl.hpp>
+
+#include <memory>
+
+// need to interface this into cmake variable (or only use this flag when it is necessary)
+//#define BLOCKTRIDICONTAINER_ENABLE_PROFILE
+#undef  BLOCKTRIDICONTAINER_ENABLE_PROFILE
+#if defined(KOKKOS_ENABLE_CUDA) && defined(BLOCKTRIDICONTAINER_ENABLE_PROFILE)
+#include "cuda_profiler_api.h"
+#endif
+
+namespace Ifpack2 {
+
+  namespace BlockTriDiContainerDetails {
+    ///
+    /// view decorators for unmanaged and const memory
+    ///
+    using do_not_initialize_tag = Kokkos::ViewAllocateWithoutInitializing;
+
+    template <typename MemoryTraitsType, Kokkos::MemoryTraitsFlags flag>
+    using MemoryTraits = Kokkos::MemoryTraits<MemoryTraitsType::Unmanaged |
+                                              MemoryTraitsType::RandomAccess |
+                                              MemoryTraitsType::Atomic |
+                                              flag>;
+    
+    template <typename ViewType>
+    using Unmanaged = Kokkos::View<typename ViewType::data_type,
+                                   typename ViewType::array_layout,
+                                   typename ViewType::device_type,
+                                   MemoryTraits<typename ViewType::memory_traits,
+                                                Kokkos::Unmanaged> >;
+    template <typename ViewType>
+    using Const = Kokkos::View<typename ViewType::const_data_type, 
+                               typename ViewType::array_layout,
+                               typename ViewType::device_type, 
+                               typename ViewType::memory_traits>;
+    template <typename ViewType>
+    using ConstUnmanaged = Const<Unmanaged<ViewType> >;    
+
+    ///
+    /// cuda specialization
+    ///
+    template<typename T> struct is_cuda        { enum : bool { value = false }; };
+#if defined(KOKKOS_ENABLE_CUDA)
+    template<> struct is_cuda<Kokkos::Cuda>    { enum : bool { value = true  }; };
+#endif
+
+    ///
+    /// utility functions
+    ///
+    template<typename CommPtrType>
+    std::string get_msg_prefix (const CommPtrType &comm) {
+      const auto rank = comm->getRank();
+      const auto nranks = comm->getSize();
+      std::stringstream ss;
+      ss << "Rank " << rank << " of " << nranks << ": ";
+      return ss.str();
+    }
+
+    ///
+    /// custom multiple varilable reduce and scan
+    ///
+    template<typename T, int N>
+    struct ArrayValueType {
+      T v[N];
+      KOKKOS_INLINE_FUNCTION 
+      ArrayValueType() {
+        for (int i=0;i<N;++i)
+          this->v[i] = 0;
+      }
+      KOKKOS_INLINE_FUNCTION 
+      ArrayValueType(const ArrayValueType &b) {
+        for (int i=0;i<N;++i) 
+          this->v[i] = b.v[i];
+      }      
+    };
+    template<typename T, int N>
+    static 
+    KOKKOS_INLINE_FUNCTION 
+    volatile ArrayValueType<T,N>& 
```

Comment:
Does Kokkos still need `volatile` overloads?  Last time I wrote a custom reduction with a value time, I didn't need them.  Not sure about array types, though.

Reply:
Without volatile overloads, it causes a compile error.